xdsl/transforms/loop_invariant_code_motion.py

```diff
@@ -32,9 +41,82 @@ def can_be_hoisted(op: Operation, target_region: Region) -> bool | None:
     for child in op.walk():
         for operand in child.operands:
             operand_owner = operand.owner
-            assert operand_owner is not None
             if op.is_ancestor(operand_owner):
                 continue
             if target_region.is_ancestor(operand_owner):
                 return False
     return True
+
+
+def _move_loop_invariant_code(
+    regions: Sequence[Region],
+    should_move_out_of_region: Callable[[Operation, Region], bool],
+    move_out_of_region: Callable[[Operation, Region], None],
+):
+    num_moved = 0
+    for region in regions:
+        # add top-level operations in the loop body to the worklist
+        worklist = [op for block in region.blocks for op in block.ops]
```

Comment:
Can we reuse `Worklist` from `pattern_rewriter` which I think does this more efficiently?

Reply:
I don't think that Worklist would be more efficient here, but I replaced this with a deque, which is probably better than a list.